A workshop report from the May 2015 non-model RNAseq workshop at UC Davis
#########################################################################

:author: C\. Titus Brown
:tags: workshops,rnaseq,davis
:date: 2015-05-06
:slug: 2015-may-workshop-report
:category: teaching

We just finished teaching the second of my RNAseq workshops at UC
Davis -- the fifth workshop I've hosted since I took a faculty
position here in VetMed.  In order, we've done a Train the Trainers, a
Data Carpentry, a reference-guided RNAseq assembly workshop, a mothur
(microbial ecology) workshop, and a de novo RNAseq assembly workshop
-- you can see all of the links at the `Data Intensive Biology
Training Program Web site
<http://dib-training.readthedocs.org/en/pub/>`__.  This workshop was
the `May de novo mRNAseq assembly workshop
<http://2015-may-nonmodel.readthedocs.org/en/dev/>`__, which I
co-taught with Tamer Mansour and Camille Scott.

The workshops are still maturing, and I'm trying to figure out how
to keep this going for the medium term, but so far I think we're
doing an OK job.  We can always improve the material and the delivery,
but I think at least we're on a good trajectory.

This workshop (and the many excellent questions raised by the
attendees) reminded me how much of RNAseq analysis is still *research*
-- it's not just a question of what assembler and quantification
method to use, but much more fundamental questions of data evaluation,
assembly evaluation, and how to tie this into the biology you're
trying to do.  My lab works on this a lot, and too much of the time we
have to say "we just don't know" - often because the experts don't
agree, or because the answer is just unknown.

I also despair sometimes that the energy and effort we're putting into
this isn't enough.  There is a huge demand, and these two day
workshops are at best a stopgap measure, and I really have no idea
whether they're going to help biologists starting from scratch to
analyze their own data.

I do have other arrows in my quiver. Once my lab "lands" at Davis
(sometime between June and September) I expect to start up a biology
"data space" of some sort, where every week people who have been
through one of my workshops can come work on their data analysis; the
hope is that, much like the `Davis R Users Group
<http://www.noamross.net/davis-r-users-group.html>`__, we can start to
build a community around biological data analysis.  Stay tuned.

I'm also planning to start running more advanced workshops.  One great
idea that Tamer pitched to me this morning was to run a follow-on
workshop entitled "publishing your transcriptome", which would focus
on quality measures and analysis downstream of your first-blush
transcriptome assembly/annotation/quantification.  I'm also hoping to
put together an "automation and reproducibility" workshop in the fall,
along with a variety of more focused workshops on specific platforms and
questions.

And, of course, we'll continue running the intro workshops.  In
addition to the mRNASeq workshops, in the fall I'd like to do
workshops on microbial genome assembly and annotation, metagenome and
metatranscriptome assembly, and advanced UCSC genome browser
use/misuse (think assembly hubs etc.).

--titus
